README
Catalyst-View-Template-Declare

The README is used to introduce the module and provide instructions on
how to install the module, any machine dependencies it may have (for
example C compilers and installed libraries) and any other information
that should be provided before the module is installed.

A README file is required for CPAN modules since CPAN extracts the README
file from a module distribution so that people browsing the archive
can use it get an idea of the modules uses. It is usually a good idea
to provide version information here so that people can decide whether
fixes for the module are worth downloading.

INSTALLATION

To install this module, run the following commands:

    perl Makefile.PL
    make
    make test
    make install


SUPPORT AND DOCUMENTATION

After installing, you can find documentation for this module with the perldoc command.

    perldoc Catalyst::View::Template::Declare

You can also look for information at:

    Search CPAN
        http://search.cpan.org/dist/Catalyst-View-Template-Declare

    CPAN Request Tracker:
        http://rt.cpan.org/NoAuth/Bugs.html?Dist=Catalyst-View-Template-Declare

    AnnoCPAN, annotated CPAN documentation:
        http://annocpan.org/dist/Catalyst-View-Template-Declare

    CPAN Ratings:
        http://cpanratings.perl.org/d/Catalyst-View-Template-Declare

COPYRIGHT AND LICENCE

Copyright (C) 2007 Jonathan Rockway

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.


Changes
Revision history for Catalyst-View-Template-Declare

0.03    14 September 2007
        - use a smarter way of finding the Catalyst context
        - strip off leading newlines (so that XML is valid)

0.02    18 August 2007
        - Convert to Catalyst::View::Templated API, deprecate
          templates in subclass

0.01    9 July 2007
        - Require T::D::Tags again (Catalyst pre-loading issue)
        - Helper generates valid code now

0.00_03 8 April 2007
        - No longer require "use Template::Declare::Tags"
        - Add a Catalyst Helper

0.00_02 7 April 2007
        - Remove duplicate output
        - Allow templates in subfiles (MyApp::View::TD::*)
        - Auto-load those subfiles

0.00_01 7 April 2007
        First version, released on an unsuspecting world.


t/lib/TestApp.pm
#!perl

package TestApp;
use strict;
use warnings;
use Catalyst;

TestApp->setup;

1;


t/lib/TestApp/View/TD.pm
package TestApp::View::TD;
use strict;
use warnings;
use base 'Catalyst::View::Template::Declare';
use Template::Declare::Tags;
template action_name => sub {
    p { "This is the action_name template." };
};

template stash => sub {
    p { "Hello, ". c->stash->{world} };
};

1;


t/lib/TestApp/View/TD/Main.pm
# Copyright (c) 2007 Jonathan Rockway <jrockway@cpan.org>

package TestApp::View::TD::Main;
use strict;
use warnings;
use Template::Declare::Tags;

template action_name => sub {
    p { "This is the action_name template." };
};

template stash => sub {
    p { "Hello, ". c->stash->{world} };
};

template methods => sub {
    p { c->hello_world };
};

1;


lib/Catalyst/View/Template/Declare.pm
package Catalyst::View::Template::Declare;
use strict;
use warnings;
use base qw(Catalyst::View::Templated);
use Class::C3;
require Module::Pluggable::Object;

our $VERSION = '0.03';

sub COMPONENT {
    my $self  = shift;
    my $c     = shift;
    my $class = ref $self || $self;
    
    # find sub-templates
    my $mpo = Module::Pluggable::Object->new(require     => 0,
                                             search_path => $class,
                                            );
    
    # load sub-templates (and do a bit of magic niceness)
    my @extras = $mpo->plugins;
    foreach my $extra (@extras) {
        $c->log->info("Loading subtemplate $extra");
        
        # load module
        if (!eval "require $extra"){
            die "Couldn't include $extra: $@";
        }
        
        # make the templates a subclass of TD (required by TD)
        {
            no strict 'refs';
            push @{$extra. "::ISA"}, 'Template::Declare';
        }
        
    }
    
    # init Template::Declare
    Template::Declare->init(roots => [$class, @extras]);
    
    # init superclasses
    $self->next::method($c, @_);
}

sub _render {
    my ($self, $template, $stash, $args) = @_;

    Template::Declare->new_buffer_frame;
    my $out = Template::Declare->show($template);
    Template::Declare->end_buffer_frame;

    $out =~ s/^\n+//g; # kill leading newlines
    return $out;
}

package c;
use PadWalker qw(peek_my);
our $AUTOLOAD;
sub AUTOLOAD {
    shift; # kill class

    # walk up the stack looking for the Catalyst context
    # in a lexical somewhere (evil, yes.)
    my $frames_up = 1;
    my $context;
    while($frames_up < 300 && !$context){
        ($context) = 
          map { $$_ } 
            grep {eval{$$_->isa('Catalyst')}} 
              values %{peek_my($frames_up++)};
    }
    die "INTERNAL ERROR: No Catalyst context found!" if !$context;
    
    $AUTOLOAD =~ s/^c:://; # kill package c
    return $context->$AUTOLOAD(@_);
}

1;
__END__

=head1 NAME

Catalyst::View::Template::Declare - Use Template::Declare with Catalyst

=head1 VERSION

Version 0.02

=head1 SYNOPSIS

Create the view:

     myapp_create.pl view TD Template::Declare

Add templates in C<< MyApp::View::TD::<name> >>:

     package MyApp::View::TD::Test;
     use Template::Declare::Tags;

     template foo => sub { html {} };
     template bar => sub {   ...   };
     1;

Then use the templates from your application:

     $c->view('TD')->template('foo');
     $c->detach('View::TD');

You can get at the Catalyst context via the C<<c>> package:

     template foo => sub { "This is the ". c->action. " action." };
     template bar => sub { "Hello, ". c->stash->{world} };

Have fun.  This is all somewhat experimental and subject to change.

=head1 DESCRIPTION

Make a view:

    package MyApp::View::TD;
    use base 'Catalyst::View::Template::Declare';
    1;

Make a template:

    package MyApp::View::TD::Root;
    use Template::Declare::Tags;
     
    template foo => sub { 
        html { 
            head { title { c->stash->{title} } };
            body { "Hello, world" }
          }
    };

In your app:

    $c->view('TD')->template('foo');
    $c->stash(title => 'test');
    $c->detach('View::TD');

And get the output:

    <html><head><title>test</title></head><body>Hello, world</body></html>

You can spread your templates out over multiple files.  If your
view is called MyApp::View::TD, then everything in MyApp::View::TD::*
will be included and templates declared in those files will be available
as though they were declared in your main view class. 

Example:

    package MyApp::View::TD::Foo;
    use Template::Declare::Tags;
    template bar => sub { ... };
    1;

Then you can set C<< $c->view('TD')->template('bar') >> and everything
will work as you expect.

=head1 METHODS

=head2 process

Render the template in C<< $self->template >>; see
L<Catalyst::View::Templated> for information on how to specify the
template.

=head2 render($template)

Render the template named by C<$template> and return the output.

=head2 COMPONENT

Not for you.

=head1 AUTHOR

Jonathan Rockway, C<< <jrockway at cpan.org> >>

=head1 BUGS

Please report any bugs or feature requests to
C<bug-catalyst-view-template-declare at rt.cpan.org>, or through the web interface at
L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Catalyst-View-Template-Declare>.
I will be notified, and then you'll automatically be notified of progress on
your bug as I make changes.

=head1 SUPPORT

Visit #catalyst on irc.perl.org, submit an RT ticket, or send me an e-mail.

The git repository is at L<http://git.jrock.us/>.

=head1 SEE ALSO

L<Catalyst::View::Templated>

=head1 ACKNOWLEDGEMENTS

L<Template::Declare>

L<Jifty>

=head1 COPYRIGHT & LICENSE

Copyright 2007 Jonathan Rockway, all rights reserved.

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.


